services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionSyncClient.java

```diff
@@ -15,80 +15,86 @@
+            if (exception.statusCode() == REDIRECT_STATUS_CODE) {
+                updateCacheFromRedirectException(exception, bucketName);
+                return operation.apply(
+                    requestWithDecoratedEndpointProvider(
+                        request,
+                        () -> bucketToRegionCache.computeIfAbsent(bucketName, this::fetchBucketRegion),
+                        serviceClientConfiguration().endpointProvider().get()));
             }
+            throw exception;
         }
-
-        return operation.apply(request);
-    }
-
-    private void handleOperationFailure(Throwable t, String bucket) {
-        //TODO: handle failure case
-    }
-
-    @SuppressWarnings("unchecked")
-    private <T extends S3Request> T requestWithDecoratedEndpointProvider(T request, String bucket) {
-        return (T) request.toBuilder()
-                          .overrideConfiguration(getOrCreateConfigWithEndpointProvider(request, bucket))
-                          .build();
     }
 
-    //TODO: optimize shared sync/async code
-    private AwsRequestOverrideConfiguration getOrCreateConfigWithEndpointProvider(S3Request request, String bucket) {
-        AwsRequestOverrideConfiguration requestOverrideConfig =
-            request.overrideConfiguration().orElseGet(() -> AwsRequestOverrideConfiguration.builder().build());
-
-        S3EndpointProvider delegateEndpointProvider = (S3EndpointProvider)
-            requestOverrideConfig.endpointProvider().orElseGet(() -> serviceClientConfiguration().endpointProvider().get());
-
-        return requestOverrideConfig.toBuilder()
-                                    .endpointProvider(BucketEndpointProvider.create(delegateEndpointProvider, bucket))
-                                    .build();
+    private void updateCacheFromRedirectException(S3Exception exception, String bucketName) {
+        Optional<String> regionStr = getBucketRegionFromException(exception);
+        // If redirected, clear previous values due to region change.
+        bucketToRegionCache.remove(bucketName);
+        regionStr.ifPresent(region -> bucketToRegionCache.put(bucketName, Region.of(region)));
     }
 
-    static final class BucketEndpointProvider implements S3EndpointProvider {
-        private final S3EndpointProvider delegate;
-        private final String bucket;
-
-        private BucketEndpointProvider(S3EndpointProvider delegate, String bucket) {
-            this.delegate = delegate;
-            this.bucket = bucket;
+    private Region fetchBucketRegion(String bucketName) {
+        try {
+            ((S3Client) delegate()).headBucket(HeadBucketRequest.builder().bucket(bucketName).build());
+        } catch (S3Exception exception) {
+            if (exception.statusCode() == REDIRECT_STATUS_CODE) {
+                return Region.of(getBucketRegionFromException(exception).orElseThrow(() -> exception));
+            }
+            throw exception;
         }
+        return null;
```

Comment:
Should we return current region instead of null if headBucket call doesn't throw exception?

Reply:
No its would be not required , since in the BucketEndpointProvider we do  a `crossRegion != null` check. It its null then we use the region passed in the EndpointParams which will be the default params